library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Closure.hs

```diff
@@ -0,0 +1,149 @@
+mallocEnvironment length =
+  -- is this the correct offset to get it to compile
+  Block.malloc (Types.addressSpace * length) environmentPtr
+
+-- We will cast this when we use it!
+functionPtr :: Type.Type
+functionPtr =
+  Types.pointerOf Type.i8
+
+environmentPtr :: Type.Type
+environmentPtr =
+  Types.pointerOf (Types.pointerOf Type.i8)
+
+-- We want a pointer of pointers to simulate an array of pointers
+environmentPtrDeref :: Type.Type
+environmentPtrDeref =
+  (Types.pointerOf Type.i8)
+
+--------------------------------------------------------------------------------
+-- Pointer Storage
+--------------------------------------------------------------------------------
+
+-- | @storeFunctionPtr@ takes the location of a closure and the
+-- function pointer to store in it. The function is bitcasted to the
+-- proper type before being stored
+storeFunctionPtr :: Types.RetInstruction m => AST.Operand -> AST.Operand -> m ()
+storeFunctionPtr location value = do
+  fnLocation <- getFunctionPtr location
+  -- cast the value for the caller
+  lamBitCasted <- Block.bitCast value functionPtr
+  Block.store fnLocation lamBitCasted
+
+-- | @storeEnvironmentPtr@ takes the location of a closure and the
+-- environment closure to store
+storeEnvironmentPtr :: Types.RetInstruction m => AST.Operand -> AST.Operand -> m ()
+storeEnvironmentPtr location value = do
+  envLocation <- getEnvironmentPtr location
+  Block.store envLocation value
+
+--------------------------------------------------------------------------------
+-- Environment Retrieval functions
+--------------------------------------------------------------------------------
+
+getFunctionPtr :: Types.RetInstruction m => AST.Operand -> m AST.Operand
+getFunctionPtr location =
+  Block.getElementPtr $
+    Types.Minimal
+      { type' = Types.pointerOf functionPtr,
+        address' = location,
+        indincies' = Block.constant32List [0, 0]
+      }
+
+loadFunctionPtr ::
+  Types.RetInstruction m => AST.Operand -> Type.Type -> m AST.Operand
+loadFunctionPtr location typ = do
+  functionPtr <-
+    Block.loadElementPtr $
+      Types.Minimal
+        { type' = functionPtr,
+          address' = location,
+          indincies' = Block.constant32List [0, 0]
+        }
+  Block.bitCast functionPtr typ
+
+getEnvironmentPtr :: Types.RetInstruction m => AST.Operand -> m AST.Operand
+getEnvironmentPtr location =
+  Block.getElementPtr $
+    Types.Minimal
+      { type' = Types.pointerOf environmentPtr,
+        address' = location,
+        indincies' = Block.constant32List [0, 1]
+      }
+
+-- | @loadEnvironmentPtr@ Loads and Dereferences the given closure
+-- environment in the given @location@
+loadEnvironmentPtr :: Types.RetInstruction m => AST.Operand -> m AST.Operand
+loadEnvironmentPtr location =
+  Block.loadElementPtr $
+    Types.Minimal
+      { type' = environmentPtr,
+        address' = location,
+        indincies' = Block.constant32List [0, 1]
+      }
```

Comment:
The case of `Block.loadElementPtr $ Types.Minimal { ...Block.constant32List...}` looks like it might be common; would it be worth factoring out a utility routine for that?

Reply:
Might be indeed, I just left it as two separate things, as what if you want an i64 index list? (silly, but I left it open for us to remove it later I guess)

Reply:
Indeed it might be worth doing that. I didn't to leave open the possibility for other index types, but yeah probably warrented